new tutorial article (#130)
junobuild/docs: blog/2024-04-30-How-to-build-a-secure-and-decentralized-blog-website-on-the-blockchain.md

---
slug: How to Build a Secure and Decentralized Blog Website on the Blockchain
title: How to Build a Secure and Decentralized Blog Website on the Blockchain
authors: [sam-the-tutor]
---

> ![](https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/ddd.1362.695.PNG)

Author: I will let you take care of it.

Unlike traditional blockchains, the Internet Computer uses a unique consensus mechanism called Threshold Relay, which allows it to achieve high transaction throughput and low latency. The platform is also designed to be highly scalable, with the ability to add more nodes and increase its computing power as demand grows. This makes the Internet Computer a promising platform for building a wide range of decentralized applications, from social media and e-commerce to finance and cloud computing.[Learn more about ICP](https://internetcomputer.org/)

## About the Project

This is a secure and decentralized blog website hosted on the blockchain using Juno. The frontend is build with Astro, which is a modern, flexible web framework focused on building fast, content-rich websites with minimal JavaScript. Here is what you will build by the end of thi article

![live project](https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/ddd.1362.695.PNG)

## Setting up the project

In this section, we will look at how to create a boilerplate template for our project using Juno

In your terminal, run the command below

```bash

> npm create juno@latest --template astro-starter

Reviewer: ```suggestion
npm create juno@latest -- --template astro-starter
```

```

> In the prompts, select `Static website or blog` as the kind of project, `Astro` as the framwork, `yes` to configure Github Actions, `no` to configure the local development emurator, `yes` to install the dependencies, and `yes` to install juno's CLI tool. Juno CLI will help us to deploy our project in the satellite.

Reviewer: Can you transfor this to a bucket list? Like

In the prompts, select:

- `Static website`...
- `No` to configue...

Also you might want to review the steps because Astro should not be asked given that we provide `--template` to the command line.

Navigate to the project folder `astro-starter` and open it in your favorite code editor.

> If every previous step is successfull, running `npm run start` will open the project in your browser and you should have something similar to this.

Reviewer: ```suggestion
If every previous step is successfull, running `npm run dev` will open the project in your browser and you should have something similar to this.
```

![template](https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/sw.1354.642.PNG)

### File Structure

Having followed the above steps, and opening the project in a code editor, your project should have a file structure similar to the one below.

![filestructure](https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/fielstruc.259.579.PNG)

> ## Project code

Reviewer: ```suggestion
## Blog code
```

> In this section,we will look at the code required for our project to function

Reviewer: ```suggestion
In this section,we will adapt the boilerplate code to transform your project into a blogging website.
```

Or something similar.

      "url": "https://tailwindcss.com/docs/installation"
    },
    {
      "title": "The Benefits of Static Site Generation",
      "image": "https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/staticsite.270.148.png",
      "description": "Static site generation offers several benefits, including improved performance, better security, and easier deployment.",
      "url": "https://www.netlify.com/blog/2016/05/02/top-ten-reasons-the-static-website-is-back/"
    }, {
        "title": "Introduction to Astro",
        "image": "https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/download.300.168.png",
        "description": "Astro is a new static site generator that makes it easy to build fast, content-focused websites.",
        "url": "https://docs.astro.build/en/getting-started/"
      }
  ]
```

This file holds our sample article data that we are using for this project.

If all the above steps are successfull, your project should look like this in the browser

![local project](https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/local.1351.693.PNG)

> ## Project Deployment

Reviewer: ```suggestion
## Deployment
```

In this section, we will look at how to deploy our project to the satellite using juno.

### Creating a satellite

We need to create a satellite that will host our project. Follow the steps below to create your own satellite.

![juno statellite](https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/createsatellite.600.338.gif)

> - Navigate to the [juno console](https://console.juno.build/) website

Reviewer: ```suggestion
- Navigate to the administration [console](https://console.juno.build/) website
```

- Login with your internet Identity
- On the dashboard, select Launch new satellite
- Provide name `myBlogSatellite` for the satellite.
- Click `Create Satellite`

> 

Reviewer: Should we add a link to Internet Identity in the info?

https://internetcomputer.org/internet-identity

Author: yep, I have added the link

> `To keep the satellite operational, the developer pays a small fee that is used to purchase the necessary cycles for the satellite's storage and computation requirements.` [Learn more about pricing](https://juno.build/docs/pricing)

Reviewer: Can you try Docusaurus adminitions?

```suggestion
:::info

To keep the satellite operational, the developer pays a small fee that is used to purchase the necessary cycles for the satellite's storage and computation requirements. [Learn more about pricing](https://juno.build/docs/pricing)

:::
```

## Connect Project to the Satellite

We need to link our project to the satellite. follow the steps below

In the project terminal, run the command `juno init` and follow the prompts

> Select `yes` to login and authorize the terminal to access your satellite in your browser, select `myBlogSatellite` as the satellite to connect the project to, `dist` as the location of the compiled app files, and `JavaScript` as the configuration file format.

Reviewer: ```suggestion
Select `yes` to login and authorize the terminal to access your satellite in your browser, select `myBlogSatellite` as the satellite to connect the project to, `dist` as the location of the compiled app files, and `TypeScript` as the configuration file format.
```

> Select `yes` to login and authorize the terminal to access your satellite in your browser, select `myBlogSatellite` as the satellite to connect the project to, `dist` as the location of the compiled app files, and `JavaScript` as the configuration file format.

Reviewer: Maybe a bucket list too?

Select:

- `Yes` to ....
- This
- That

Author: I will work on that.

> If the above step is successful, a new file `juno.config.js`

Reviewer: ```suggestion
If the above step is successful, a new file `juno.config.ts`
```

 will be added at the root of our project folder. It contains the configuration necessary for our poject to connect to the satellite. You need this file if your project is to be deployed successfully to the satellite. Learn more about this [configuration](https://juno.build/docs/miscellaneous/configuration)

### Compiling and deploying the Project

 Now that we connected our project to the satellite, we have to compile and deploy project to the satellite

 ```bash
 npm run build
 ```

 The above command compiles our project and outputs the compiled files in the `dist` folder

 ```bash
juno deploy
 ```

This will deploy our compiled files to the satellite that we connected linked our project to.

At this stage, if all the previous steps are successful, `juno deploy` command will output a link whixh is in this format `https://<SATELLITE_ID>.icp0.io` where `SATELLITE_ID` is the id of the satellite that we connected our project to.

Opening the link in the browser, you should have something like this below
![live project](https://a2ede-rqaaa-aaaal-ai6sq-cai.raw.icp0.io/uploads/ddd.1362.695.PNG)

> 

Reviewer: Here we can maybe add:

```
:::tip

Running `juno open` in your terminal opens your project in your favorite browser.

:::
```

WDYT?

If you have reached this step, well done, you have successfully deployed your first blog website on the blockchain using Juno.

## Setting up Github Actions

If you noticed in the previous steps, every time we make changes to our project, we have to manually run the commands that compile and deploy our code to the satellite. But in this section, we will learn how to automate these tasks using [Gihtub Actions](https://docs.github.com/en/actions) so that whenever we make changes to our project, these changes are automatically deployed to oour satellite

> In our project, we have a folder `.github` which contains the file `deploy.yaml`. This file has all the configurations required to setup Github Actions in our project. This folder must be present in your poject to successfully setup Github Actions. You can add it manually if you dont have it in your project. Below are the contents of the `deploy.yaml` file

Reviewer: where is the content of the `deploy.yaml`?

```bash
git init
git remote add origin https://github.com/sam-thetutor/myfirstBlog.git
git add .
git commit -m "my first commit"
git push -u origin main
```

The above code established the required connection to our remote Github repo, and pushes all our project code to that repo.Now every time we make changes to our project, all we have to do is push these changes to our github repo and they will be deployed to our satellite automatically. [Learn more about setting up Github Actions with Juno](https://juno.build/docs/guides/github-actions)

## Next Steps

Now that we have successfully hosted our blog website on the blockchain, you can go ahead and add more articles to the blog to showcase your skills. You can also add more features on the website to make it more robust.

## Conclusion

In this article, we have looked at how to create a boilerplate template project using juno, how to create a satellite from the juno console, writing code for our project, how to connect the satellite to the our local project, deploying our project to the satellite and configuring Github Actions to automate compiling and deployment tasks for our project

> 

Reviewer: You can add such a footer at the end of the post if you like it? Don't hesitate to review it.

```suggestion
---

👋

Stay connected with Juno by following us on [Twitter](https://twitter.com/junobuild) to keep up with our latest updates.

And if you made it this far, we’d love to have you join the Juno community on [Discord](https://discord.gg/wHZ57Z2RAG). 😉
```
